scipy/special/boost_special_functions.h

```diff
@@ -760,46 +760,94 @@ ncf_pdf_double(double x, double v1, double v2, double l)
 
 template<typename Real>
 Real
-ncf_cdf_wrap(const Real x, const Real v1, const Real v2, const Real l)
+ncf_cdf_wrap(const Real v1, const Real v2, const Real l, const Real x)
 {
-    if (std::isfinite(x)) {
-        return boost::math::cdf(
-            boost::math::non_central_f_distribution<Real, StatsPolicy>(v1, v2, l), x);
+    if (std::isnan(x) || std::isnan(v1) || std::isnan(v2) || std::isnan(l)) {
+	return NAN;
     }
-    // -inf => 0, inf => 1
-    return 1.0 - std::signbit(x);
+    if ((v1 <= 0) || (v2 <= 0) || (l <= 0)) {
+	sf_error("ncfdtr", SF_ERROR_DOMAIN, NULL);
+	return NAN;
+    }
+    if (std::isinf(x)) {
+	// -inf => 0, inf => 1
+	return 1.0 - std::signbit(x);
+    }
+    Real y;
+    try {
+	/* Based on a study of the source code for the cdf of the noncentral F
+	 * distribution within Boost at:
+	 * https://github.com/boostorg/math/blob/develop/include/boost/math/distributions/non_central_f.hpp
+	 * which calls the cdf of the noncentral Beta distribution:
+	 * https://github.com/boostorg/math/blob/develop/include/boost/math/distributions/non_central_beta.hpp
+	 * It does not appear that Boost can raise any errors here other than domain
+	 * error, but domain errors are already handled earlier in the wrapper. If
+	 * this try-catch block catches something, that would be unexpected, but it
+	 * feels wrong to call to Boost directly without a try-catch block when
+	 * SpecialPolicy is being used.
+	 */
+	y = boost::math::cdf(
+                boost::math::non_central_f_distribution<Real, SpecialPolicy>(v1, v2, l), x);
+    } catch (...) {
+        sf_error("ncfdtr", SF_ERROR_OTHER, NULL);
+        y = NAN;
```

Comment:
I've discovered there are some cases that the cdflib implementation can kind of handle that the new one cannot. They are extreme cases though, one is

```
ncfdtr(1e-100, 3, 1.5, 1e100)
# on main: np.float64(0.611813014416088)
# on this branch: np.float64(nan)

# mpmath reference implementation discussed earlier
mp_ncfdtr(1e-100, 3, 1.5, 1e100)
# 0.611815287345399
```

With a boost math evaluation error getting caught in this catch-all here. Ideally what we could do is fall back to the cdflib implementation here (and in the similar catch-all in `ncfdtri`) and set `SF_ERROR_LOSS` signaling reduced precision. Right now the cdflib implementation is defined in ` .pxd` file and I think it would be difficult to make this work. The good news is that I've already translated most of cdflib into C++ headers, but just haven't PR'd any of that work, and then we could just include the relevant header here, or restructure things in some other way. I think I might be OK with temporarily return NAN for such extreme cases until enough of my cdflib translation work lands.

Reply:
I think the NaNs are preferable to inaccurate results here.

Reply:
At least for this case, we should submit a bug report for Boost. For small values of the noncentrality parameter, there should be no trouble getting accurate results by directly evaluating the CDF from the definition unless `betainc` is inaccurate, and it's not in this case here.